File: java-example/src/test/groovy/no/javatec/pubsubemulator/spock/pubsub/testinfra/TopicAndSubscriptionGenerator.java
package no.javatec.pubsubemulator.spock.pubsub.testinfra;

import io.micronaut.context.annotation.Requires;
import no.javatec.pubsubemulator.spock.configuration.GcpConfigProperties;
import no.javatec.pubsubemulator.spock.configuration.PubSubConfigProperties;
import com.google.api.gax.rpc.AlreadyExistsException;
import com.google.cloud.pubsub.v1.SubscriptionAdminClient;
import com.google.cloud.pubsub.v1.TopicAdminClient;
import com.google.pubsub.v1.PushConfig;
import com.google.pubsub.v1.SubscriptionName;
import com.google.pubsub.v1.TopicName;
import jakarta.inject.Singleton;

import static no.javatec.pubsubemulator.spock.CustomEnvironment.PUBSUB_CONFIG;

@Singleton
@Requires(env = {PUBSUB_CONFIG})
public class TopicAndSubscriptionGenerator {

    private final TopicAdminClient topicAdminClient;
    private final SubscriptionAdminClient subscriptionAdminClient;
    private final GcpConfigProperties gcpConfigProperties;
    private final PubSubConfigProperties pubSubConfigProperties;

    public TopicAndSubscriptionGenerator(
            TopicAdminClient topicAdminClient,
            SubscriptionAdminClient subscriptionAdminClient,
            GcpConfigProperties gcpConfigProperties,
            PubSubConfigProperties pubSubConfigProperties) {

        this.topicAdminClient = topicAdminClient;
        this.subscriptionAdminClient = subscriptionAdminClient;
        this.gcpConfigProperties = gcpConfigProperties;
        this.pubSubConfigProperties = pubSubConfigProperties;
    }

    public void createTopicAndSubscriptions() {
        final var topic = TopicName.of(gcpConfigProperties.projectId(), pubSubConfigProperties.topic());
        createTopic(topic);
        createSubscription(topic, pubSubConfigProperties.subscription());
    }

    private void createSubscription(TopicName topic, String subscriptionName) {
        final var subscription = SubscriptionName.of(gcpConfigProperties.projectId(), subscriptionName);

        try {
            subscriptionAdminClient.createSubscription(
                    subscription,
                    topic,
                    PushConfig.getDefaultInstance(),
                    100);
        } catch (AlreadyExistsException e) {
            // this is fine, already created
            subscriptionAdminClient.getSubscription(subscription);
        }
    }

    private void createTopic(TopicName topic) {
        try {
            topicAdminClient.createTopic(topic);
        } catch (AlreadyExistsException e) {
            // this is fine, already created
            topicAdminClient.getTopic(topic);
        }
    }
}

File: kotlin-example/src/test/kotlin/no/javatec/pubsubemulator/kotest/pubsub/testinfra/TopicAndSubscriptionGenerator.kt
package no.javatec.pubsubemulator.kotest.pubsub.testinfra

import com.google.api.gax.rpc.AlreadyExistsException
import com.google.cloud.pubsub.v1.SubscriptionAdminClient
import com.google.cloud.pubsub.v1.TopicAdminClient
import com.google.pubsub.v1.Subscription
import com.google.pubsub.v1.SubscriptionName
import com.google.pubsub.v1.TopicName
import io.micronaut.context.annotation.Requires
import io.micronaut.context.annotation.Value
import jakarta.inject.Singleton
import no.javatec.pubsubemulator.kotest.CustomEnvironment.PUBSUB_CONFIG

@Singleton
@Requires(env = [PUBSUB_CONFIG])
class TopicAndSubscriptionGenerator(
    private val topicAdminClient: TopicAdminClient,
    private val subscriptionAdminClient: SubscriptionAdminClient,
    @Value("\${gcp.project-id}") private val projectId: String,
    @Value("\${pubsub.topic}") private val topicName: String,
    @Value("\${pubsub.subscription}") private val subscriptionName: String
) {
    fun createTopicAndSubscriptions() {
        val topic = TopicName.of(projectId, topicName)
        createTopic(topic)
        createSubscription(topic, subscriptionName)
    }

    private fun createSubscription(
        topic: TopicName,
        subscriptionName: String
    ) {
        val subscription = SubscriptionName.of(projectId, subscriptionName)
        try {
            subscriptionAdminClient.createSubscription(Subscription.newBuilder()
                .setName(subscription.toString())
                .setTopic(topic.toString())
                .setEnableMessageOrdering(true)
                .setAckDeadlineSeconds(100)
                .build()
            )
        } catch (e: AlreadyExistsException) {
            // this is fine, already created
            subscriptionAdminClient.getSubscription(subscription)
        }
    }

    private fun createTopic(topic: TopicName) {
        try {
            topicAdminClient.createTopic(topic)
        } catch (e: AlreadyExistsException) {
            // this is fine, already created
            topicAdminClient.getTopic(topic)
        }
    }
}